tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,209 @@
+import random
+from collections import defaultdict
+from itertools import islice
+from typing import Dict, List, Optional, Sequence, Union
+
+import numpy as np
+import torch
+
+from tango.common.dataset_dict import DatasetDictBase
+from tango.common.exceptions import ConfigurationError
+from tango.common.lazy import Lazy
+from tango.common.tqdm import Tqdm
+from tango.format import Format, JsonFormat
+from tango.step import Step
+
+from .data import DataLoader
+from .eval_callback import EvalCallback
+from .model import Model
+from .util import check_dataset, move_to_device
+
+
+@Step.register("torch::eval")
+class TorchEvalStep(Step):
+    """
+    A basic PyTorch evaluation loop that pairs well with :class:`TorchTrainStep`.
+
+    .. tip::
+
+        Registered as a :class:`~tango.step.Step` under the name "torch::eval".
+
+    .. warning::
+
+        By default the metrics specified by the ``metric_names`` parameter
+        are aggregated by simply averaging across batches.
+        This behavior is usually correct for metrics like "loss" or "accuracy",
+        for example, but may not be correct for other metrics like "F1".
+
+        If this is not correct for your metric you will need to handle the aggregation
+        internally in your model or with an :class:`EvalCallback`
+        using the :meth:`EvalCallback.post_batch()` method.
+        Then set the parameter ``auto_aggregate_metrics`` to ``False``.
+
+    """
+
+    DETERMINISTIC = True
+    CACHEABLE = True
+    FORMAT: Format = JsonFormat()
+
+    def run(  # type: ignore[override]
+        self,
+        model: Model,
+        dataset_dict: DatasetDictBase,
+        dataloader: Lazy[DataLoader],
+        test_split: str = "test",
+        seed: int = 42,
+        eval_steps: Optional[int] = None,
+        log_every: int = 1,
+        device: Optional[Union[int, str, torch.device]] = None,
```

Comment:
Do we expect different results when the eval is run on a different device? I know we had this discussion for training. Does that also apply to evaluations?

Do we expect different results if the step runs on device 1 vs. device 2? Because that's how Tango will treat this.

Reply:
Ideally it shouldn't matter, but it might (CPU vs GPU). But if you leave `device` as `None`, tango will automatically decide between GPU and CPU, using GPU whenever there is one available. I don't know if there's a reliable way to choose *which* GPU to use when there is more than one available.

Reply:
If we assume that CPU and GPU training are different, we can't also auto-detect whether we want GPU support or not. Or at least, the auto-detection has to happen before the step gets instantiated. Otherwise, when you run with `device=None`, you'll get different results depending on which machine you run on, but the same unique id.

If we assume that CPU and GPU training are the same, we should not have a `device` parameter at all.

I think we can assume that running on one GPU is the same as running on another. I know this is not strictly true, but it's true enough.

I would argue that the difference between CPU and GPU also falls into this category: Not strictly the same, but same enough for practical purposes. But I'm willing to be convinced.

Reply:
The way it is right now leaves this decision up to the user. If they leave `device=None`, they are telling tango that they will get the same result on GPU or CPU. Otherwise they can explicitly `device` to "0" or "cpu", whatever they want.